[Migration Prep Required] Feature/Registration Responses Validation [ENG-874]

pattisdr opened the pull request:
❗️ Locks to the same release as ENG-842 -  this finishes adding being able to update registration_responses when updating drafts.  ENG-842 builds `registration_responses` on create, but not update. 

## Purpose

- Add validation of `registration_responses` field, which is the more flattened field storing responses to supplemental registration questions that corresponds to the more flattened schema blocks.
- Validation ensures that required fields are present (if `required_fields` is True), that multiple choice responses are one of the provided responses, and that file information is formatted properly.  Also doesn't allow any extra keys to go into the db, unless the key matches a `registration_response_key`

## Changes
- Builds jsonschema for validating `registration_responses` on schemas
- Validates `registration_responses` when a DraftRegistration is created or updated, and `registration_responses was passed in.  If old workflow (`registration_metadata`) was passed in, use existing validation.
- Validates that file information is correct on `registration_metadata`. Your filename, and sha256 have to be accurate to register.


## QA Notes
- This PR allows updating a draft to update either `registration_responses` or `registration_metadata` through the API.  If you update the new `registration_responses` field, it will be expanded and stored in `registration_metadata` (verify through the API).
- If you update the old Draft.registration_metadata field through the API, it should be flattened and likewise stored under `registration_responses`.
- Verify you can update both `registration_metadata`, and `registration_responses` on the Draft Registration serializer at the same time
- Verify that `registration_metadata` is no longer a required field when updating a draft. 
- `registration_responses` or `registration_metadata` on the DraftRegistration are validated when updating the draft, not both.  
- Verify that required fields are not required to be present when updating the draft
- Verify that you get 400 errors when trying to update `registration_responses` incorrectly. For example, your data is the wrong type, your answer choice doesn't match one of the multiple choice answers, or you include a question that doesn't exist.
- This PR is not validating when the registration is created (but this validation will be turned on in https://openscience.atlassian.net/browse/ENG-366)
-  Verify that invalid file information can't be sent in for a question. For example, the nodeId on the file has to be the target.  The `file_name` has to be the name of the file.  The sha256 has to correspond to a version of that file, otherwise, you cannot create a registration.

⭐️ This used to be caching the jsonschema on the RegistrationSchema, but that has been removed; we're calculating on the fly now.
## Documentation

<!-- Does any internal or external documentation need to be updated?
     - If the API was versioned, update the developer.osf.io changelog.
     - If changes were made to the API, link the developer.osf.io PR here.
-->

## Side Effects

<!-- Any possible side effects? -->

## Ticket

https://openscience.atlassian.net/browse/ENG-874

pattisdr committed:
Exploring migrating over Registration.registered_meta and DraftRegistration.registration_metadata to answers.

pattisdr committed:
Merge branch 'feature/ENG608-flatter-form-blocks' into feature/ENG609_registered_meta_migration_analysis

pattisdr committed:
Rebase migrations and update "answers" field to "registration_responses".

pattisdr committed:
Modify migration of registration_responses to accommodate multi-select responses and file responses.

pattisdr committed:
Turn migration of registration_responses into a management command.

pattisdr committed:
Add improvement to registration_responses migration - no guarantee that any of the registration_response keys exist in `registration_metadata/registered_meta`.

pattisdr committed:
Consolidate a lot of the functionality from the migrate_registration_responses management command into osf/utils/registrations to be used when flattening registration_metadata as part of supporting both workflows.

pattisdr committed:
Consolidate flatten_registration_metadata behavior into a mixin to be shared on the DraftRegistration and Registration model.

pattisdr committed:
Have an update to `draft.registration_metadata` likewise flatten and populate `draft.registration_responses`.  Registering a node also propagates "registration_responses" from the "registered_meta" field.

pattisdr committed:
Add expand_registration_responses to be used on the DraftRegistration and Registration model.  If `registration_responses` is supplied on the DraftRegistration, populate the `registration_metadata` as well (is an expanded version of registration_responses).

pattisdr committed:
Anonymize registration registration_responses, similar to how registered_meta is anonymized.

pattisdr committed:
Purposefully set registration_responses to be an empty dictionary in migration testing.

pattisdr committed:
When expanding registration_responses -> registratio_metadata, add in a nested field for files, that's needed for rendering in legacy FE.

pattisdr committed:
Modify files count migration to use proper contenttype.

pattisdr committed:
After migration that adds and populates existing registration_responses_migrated fields with False, alter the field so future values get a default of True.

pattisdr committed:
Add testing of veer expansion into registration_metadata to check for bugs.

pattisdr committed:
Merge branch 'feature/ENG608-flatter-form-blocks' into feature/ENG842_registered_meta_migration

pattisdr committed:
Merge branch 'feature/ENG608-flatter-form-blocks' into feature/ENG842_registered_meta_migration

pattisdr committed:
Rebase migrations.

pattisdr committed:
Add a migration that stores the jsonschema on a particular RegistrationSchema version so it doesn't have to be calculated each time.  The question's title is stored under the "description" annotation on each question to help with the error response.

Add RegistrationSchema.validate_registration_responses.

pattisdr committed:
Validate registration_responses in the API on a draft registration when passed in.  Required fields on `registration_responses` are not required to be present when validating the draft.

- Make registration_metadata not required when updating a draft (since you might be updating registration_responses instead).
- Only allow `registration_responses` or `registration_metadata` when updating the draft. Not both.

pattisdr committed:
When `ensure_schemas` is run, which updates or creates new RegistrationSchemas, update `registration_responses_jsonschema` as well.

pattisdr commented on api/nodes/serializers.py:
```diff
@@ -1497,6 +1497,32 @@ class DraftRegistrationSerializer(JSONAPISerializer):
     def get_absolute_url(self, obj):
         return obj.absolute_url
 
+    def update_metadata(self, draft, metadata, reviewer, required_fields=False):
```
Consolidates this because it was being called 2x - once on create, and once on updating a DraftRegistration

pattisdr commented on api/nodes/serializers.py:
```diff
@@ -1497,6 +1497,32 @@ class DraftRegistrationSerializer(JSONAPISerializer):
+            raise exceptions.ValidationError(
+                'You cannot include both `registration_metadata` and `registration_responses` in your request. Please use' +
+                ' `registration_responses` as `registration_metadata` will be deprecated in the future.',
+            )
```
While we support both workflows, writing to `registration_metadata` or `registration_responses` will build the corresponding data and write to the other field. So we don't want to be writing to both!

pattisdr commented on api/nodes/serializers.py:
```diff
@@ -1540,13 +1555,15 @@ class Meta:
+    (new workflow), not both.  registration_metadata will eventually be deprecated.
     """
     id = IDField(source='_id', required=True)
-    registration_metadata = ser.DictField(required=True)
```
Before, we required this field when updating a `DraftRegistration` because it was the only field you *could* update.  Now that you can update either `registration_metadata` or `registration_responses`, it doesn't make sense to make `registration_metadata` required here. (And in fact, we want to move to the new workflow and encourage writing to `registration_responses`!

pattisdr commented on api/registrations/serializers.py:
```diff
@@ -543,6 +543,8 @@ def create(self, validated_data):
+            # on the draft, the other field was built and populated as well.  Both should exist.
             draft.validate_metadata(metadata=draft.registration_metadata, reviewer=reviewer, required_fields=True)
         except ValidationValueError:
             log_exception()  # Probably indicates a bug on our end, so log to sentry
```
This area validates a draft's `registration_metadata` before attempting to create a registration, this time, asserting that all the required fields are present.   We turned off throwing an error here when we started using v2 create registrations as part of partial registrations.  Our jsonschema validation wasn't perfect, so we were getting errors that weren't our fault. 


Right now, we don't throw an error, we just log to sentry if there's an error. https://openscience.atlassian.net/browse/ENG-366 should check if we're still getting sentry errors that are all fault, and if we're not, we should restore throwing an error here.

pattisdr commented on api/registrations/serializers.py:
```diff
@@ -543,6 +543,8 @@ def create(self, validated_data):
         try:
+            # Still validating metadata, but whether `registration_responses` or `registration_metadata` were populated
+            # on the draft, the other field was built and populated as well.  Both should exist.
             draft.validate_metadata(metadata=draft.registration_metadata, reviewer=reviewer, required_fields=True)
```
We should consider validating `registration_responses` instead! (Leaving for now, to be addressed in https://openscience.atlassian.net/browse/ENG-366)

pattisdr replied:
ah, a lot of the archiving code still looking at `registration_metadata` though -

pattisdr commented on osf/migrations/0186_registrationschema_registration_responses_jsonschema.py:
```diff
@@ -0,0 +1,35 @@
+def migrate_registration_responses_jsonschema(state, schema):
+    RegistrationSchema = state.get_model('osf', 'registrationschema')
+    for schema in RegistrationSchema.objects.all():
+        schema.registration_responses_jsonschema = build_flattened_jsonschema(schema)
```
Stashes the jsonschema on each RegistrationSchema so we don't have to do extra db queries every time we need to validate `registration_responses`

pattisdr commented on osf/models/metaschema.py:
```diff
@@ -148,6 +149,43 @@ def validate_metadata(self, metadata, reviewer=False, required_fields=False):
+        """
+        validation_schema = self.registration_responses_jsonschema
+        if not required_fields and validation_schema.get('required', None):
+            del validation_schema['required']
```
We don't want to enforce the presence of required fields when the DraftRegistration is being edited, we just need all the required fields present by the time we create the registration itself.  

Here, I just delete the required bit from the jsonschema before validating.

pattisdr commented on osf/models/metaschema.py:
```diff
@@ -148,6 +149,43 @@ def validate_metadata(self, metadata, reviewer=False, required_fields=False):
+            relative_path = getattr(e, 'relative_path', None)
+            question_id = relative_path[0] if relative_path else ''
+            if properties.get(question_id, None):
+                question_title = properties.get(question_id).get('description') or question_id
```
Annotated the question title, where it existed on the `description` field on the jsonschema, so it could be used here in making a clearer error message.  Input welcome on better jsonschema fields to annotate with.  Looked like `title` was only top-level.

pattisdr committed:
Modify validation to expect `file_name` and `file_id` as file keys.

pattisdr commented on osf/utils/migrations.py:
```diff
@@ -176,6 +191,9 @@ def ensure_schemas(*args):
             logger.info('Added schema {} to the database'.format(schema['name']))
 
+        schema_obj.registration_responses_jsonschema = build_flattened_jsonschema(schema_obj)
+        schema_obj.save()
```
if RegistrationSchemas are added or updated, the cached jsonschema dict will be added/updated as well. (when `ensure_schemas` is run

pattisdr commented on osf/utils/migrations.py:
```diff
@@ -413,6 +431,92 @@ def unmap_schemablocks(*args):
+    ).values_list('display_text', flat=True)
+
+    return {
+        'enum': list(options)
```
needs to be a list for jsonschema

pattisdr committed:
Removes old `chunk_id` name from comment.

pattisdr committed:
Make some changes to migrations to get around historical migrations issues -

- Modify `ensure_schemas` so it only attempts to add to `registration_responses_jsonschema` for migrations after this field exists (this method is called on historical migrations, and we need it to work when migrating an empty db)
- Move the schema migration for adding the registration_responses_jsonschema to an earlier migration to avoid another historical migration issue (hasn't been released yet, so we can update an older migration).

pattisdr committed:
Merge branch 'feature/ENG608-flatter-form-blocks' into feature/ENG842_registered_meta_migration

pattisdr committed:
Rebase migrations and rename question-title to question-label.

pattisdr committed:
Merge branch 'feature/ENG842_registered_meta_migration' into feature/ENG874_registration_responses_validation

pattisdr committed:
Rebase migration and use `question-label` block instead of `question-title`.

pattisdr committed:
Responding to CR -

1) Query for contributors-input block_type to anonymize registration_responses with anon VOL's
2) Use .count() instead of len() to log total # of drafts/regs remaining when migrating registration_responses
3) Don't give NullBooleanFields a default value to save migration time
4) Update property name `get_registration_schema` to `registration_schema` for clarify on Registrations.  For corresponding DraftRegistration model, just using attribute.
5) Pull out the file id from the viewUrl using regex, and return whatever file information I have
6) Expect sha256 to be under "extra"

aaxelb commented on api_tests/nodes/views/test_node_draft_registration_detail.py:
```diff
@@ -433,6 +420,37 @@ def test_required_metaschema_questions_not_required_on_update(
         assert res.json['data']['attributes']['registration_metadata']['q2']['value'] == 'New response'
         assert 'q1' not in res.json['data']['attributes']['registration_metadata']
 
+    def test_required_registration_responses_questions_not_required_on_update(
```
Could we also test that an update with invalid responses gives a `400`?

pattisdr replied:
added a few more tests around this, plus there's model-level testing.  This will need a lot of QA.  Also, when we throw on validation when registrations are created, I think we're going to get a lot of test failures.

pattisdr replied:
I'm getting specific file things after turning on validation when registering a draft
```
*** ValidationError: {u'seen': True, u'user': u'b2u8h'} is not of type 'string'

Failed validating 'type' in schema['properties'][u'q7']['properties']['value']['properties'][u'uploader']['properties']['extra']['items']['properties']['data']['properties']['extra']['properties']['latestVersionSeen']:
    {'type': 'string'}

On instance[u'q7']['value'][u'uploader']['extra'][0]['data']['extra']['latestVersionSeen']:
    {u'seen': True, u'user': u'b2u8h'}
```

pattisdr replied:
Maybe updating OSF_UPLOAD_EXTRA_SCHEMA will help with ENG-366, also that schema might have to be updated to include the registration_responses things are being stashed there? just a thought, haven't dug much, but we do need to get validation turned on before we register!

aaxelb commented on api_tests/nodes/views/test_node_draft_registration_list.py:
```diff
@@ -510,6 +510,58 @@ def test_required_metaschema_questions_not_required_on_post(
         assert data['embeds']['branched_from']['data']['id'] == project_public._id
         assert data['embeds']['initiator']['data']['id'] == user._id
 
+    def test_required_registration_responses_questions_not_required_on_post(
```
ditto, test for `400` response

pattisdr replied:
done, but is going to need a lot of QA.

aaxelb reviewed (approved):
Looks good! Only suggestion I had was "more tests", which you've already said you know, so this is not very useful.

pattisdr commented:
@aaxelb, I can't find the JIRA ticket, but I just thought of something related to us passing in the `sha25`  under `registration_responses` and this not being validated.

Todd made a ticket a long time ago that was deferred (b/c we weren't using v2 for registrations), where v2 `registration_responses` file information wasn't being validated.  No enforcing that the files passed in attached to supplemental questions are actually files on the node, and how in theory, those file_ids/file names/and sha256 do not have to be correct.

Do we want to add overhead for verifying that any attached files are actually present, the file_name is actually the name on the file, and the sha256 is a version, etc. or do we trust the input?  There's a `_find_orphan_files` on v2 register that looks like it is checking at least some of this, but probably all of our bases aren't covered if people are using v2 directly.

aaxelb commented:
@pattisdr That's a good point! I think checking that the given file info is valid and correct is well worth the overhead cost -- it's ok for the last step in creating a registration to take a hair longer.

pattisdr committed:
Pull from https://github.com/CenterForOpenScience/osf.io/pull/9093.  Pass in the DraftRegistration to Node.register_node, rather than the registration_metadata.

pattisdr commented:
ok, also as an aside, the consequences of this sha256 being wrong, for example, are just that the link to the file on the registration supplemental questions won't work, we're not uploading any new files or anything like that!

I'll see if I can get some file validation added at the end

pattisdr committed:
Override contributor questions on Registration.registered_meta and Registration.registration_responses during the registration process with a comma separated list of the bibliographic contributors.

- This is to make sure there's no discrepancies between the string of contributors on registration_responses and the bibliographic contributors on the registration itself.

pattisdr committed:
Merge branch 'feature/ENG842_registered_meta_migration' into feature/ENG874_registration_responses_validation

pattisdr committed:
Add API tests for invalid updating of registration_response data.

pattisdr commented:
as an aside, @aaxelb, Archiver is still looking at files in `registration_metadata`, not `registration_responses`. We should be okay if we're reversing properly, but at some point we should shift everything to `registration_responses`

pattisdr committed:
Add sha256 to file validation.

Also allow an empty string for multiple choice responses.

pattisdr committed:
WIP commit - checking file validity on supplemental registration questions before registering.

pattisdr committed:
Piggyback off of error thrown for orphaned files when there are errors with file information:

- Don't look at registration_responses information, but rather, registration_metadata
- Look at fields that should exist whether they came in under registration_responses or registration_metadata
- nodeId, file name, and sha256 all need to be right to register.

pattisdr commented on osf/utils/migrations.py:
```diff
@@ -413,6 +435,96 @@ def unmap_schemablocks(*args):
+
+    options = list(options)
+    if '' not in options:
+        options.append('')
```
@aaxelb, this line right here is going to cause a small problem.  If a field is not required, multiple choice questions need to accept an empty string for a time. But once it's required (when we're actually registering, and it's required), an empty string should not be accepted.

Existing validation for `registration_metadata` only adds an empty string to the jsonschema if it's not required, but we have less flexibility to be dynamic about this, because we're caching these schemas.

This makes me lean against caching, unless we cache a required and a non-required version.

aaxelb replied:
Hm yeah, that makes me lean against caching too. It's easy enough to build it on the fly, so we should probably revert to that. Might be generally good to consider caching a separate optimization task from whatever we're building.

pattisdr replied:
ok I'll see if I can pull this out really quick, apologies, I just noticed this yesterday.

pattisdr commented on api/registrations/serializers.py:
```diff
@@ -567,13 +569,55 @@ def _find_orphan_files(self, registering, draft):
+            # Specified sha256 does not match a version on the specified file
+            return False
+
+        return True
```
hmm, this is going to need lots of testing, if we're validating file information before registering.

Note that I'm still using fields on `registration_metadata` here because the archiving process is still looking at these fields.  I am attempting to use fields that should always exist, regardless of whether `registration_metadata` was written to directly, or it was built from `registration_responses`. v 

A question I do have is what if the file is renamed after the file is linked in the registration responses?

aaxelb replied:
Hmm, if the file is renamed after it's linked to the draft but before the final registration... I think I'd expect `file_name` to be updated with the actual name when the registration is created? But it might make more sense to think of it as a linked file _version_, so we'd use the name and hash from when the user selected it in the draft.

Thinking about the archiving process... does the archiver update the registration's `registered_meta` (or the draft's `registered_metadata`) with the view URL of the archived file? Do we need to make it also update `registered_responses`?

pattisdr replied:
The archiver updates the registrations `registered_meta` with the viewURL of the archived file.     https://github.com/CenterForOpenScience/osf.io/blob/develop/website/archiver/utils.py#L310

`registration_metadata/registered_meta` is still pretty integral to archiver, that code hasn't been touched.  You can update `registered_responses` too, but I wouldn't rely on `registered_responses` alone unless you're updating archiver code at this point.  And even still, we need to maintain both workflows for now?

aaxelb replied:
Yeah let's not redo archiver just yet. I think we'll want to leave the logic that uses `registered_meta` to locate the linked files/hashes, check that they're present and valid, and update `registered_meta` with correct URLs; then just add a bit that *also* updates `registration_responses`.

pattisdr commented on api/registrations/serializers.py:
```diff
@@ -567,13 +569,55 @@ def _find_orphan_files(self, registering, draft):
+
         return orphan_files
 
+    def _is_attached_file_valid(self, file_metadata, registering):
```
make sure this is still being called when reconciled with DraftNode work (new API endpoints there, but they share code with old, so this should still be called).

pattisdr commented:
back to you @aaxelb, with an attempt at validating file information before registering. Also was seeing issues when trying to restore validation before registering, this is going to take some time https://openscience.atlassian.net/browse/ENG-366

pattisdr committed:
Build jsonschemas to validate `registration_responses` on the fly instead of caching, since the schemas have different data depending on whether fields are required or not.

When we're editing a draft, we're not enforcing that required fields are there, only when we're actually registering.

pattisdr commented:
@aaxelb -commit added to calculate schemas on the fly - better handles whether required multiple choice questions are validated as well.

pattisdr commented:
closing - work is folded into https://github.com/CenterForOpenScience/osf.io/pull/9174
